### PR TITLE
[SPARK-30462][SS] Streamline the logic on file stream source and sink metadata log to avoid memory issue

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

In many operations on CompactibleFileStreamLog reads a metadata log file and materializes all entries into memory. As the nature of the compact operation, CompactibleFileStreamLog may have a huge compact log file with bunch of entries included, and for now they're just monotonically increasing, which means the amount of memory to materialize also grows incrementally. This leads pressure on GC.

This patch proposes to streamline the logic on file stream source and sink whenever possible to avoid memory issue. To make this possible we have to break the existing behavior of excluding entries - now the `compactLogs` method is called with all entries, which forces us to materialize all entries into memory. This is hopefully no effect on end users, because only file stream sink has a condition to exclude entries, and the condition has been never true. (DELETE_ACTION has been never set.)

Based on the observation, this patch also changes the existing UT a bit which simulates the situation where "A" file is added, and another batch marks the "A" file as deleted. This situation simply doesn't work with the change, but as I mentioned earlier it hasn't been used. (I'm not sure the UT is from the actual run. I guess not.)

### Why are the changes needed?

The memory issue (OOME) is reported by both JIRA issue and user mailing list.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

* Existing UTs
* Manual test done

The manual test leverages the simple apps which continuously writes the file stream sink metadata log.

https://github.com/HeartSaVioR/spark-delegation-token-experiment/commit/bea7680e4c588f455f8c3181a96c9eff5002fa1a

The test is configured to have a batch metadata log file at 1.9M (10,000 entries) whereas other Spark configuration is set to the default. (compact interval = 10) The app runs as driver, and the heap memory on driver is set to 3g.

> before the patch

Screen Shot 2020-06-23 at 3 37 44 PM

It only ran for 40 mins, with the latest compact batch file size as 1.3G. The process struggled with GC, and after some struggling, it threw OOME.

> after the patch

Screen Shot 2020-06-23 at 3 53 29 PM

It sustained 2 hours run (manually stopped as it's expected to run more), with the latest compact batch file size as 2.2G. The actual memory usage didn't even go up to 1.2G, and be cleaned up soon without outstanding GC activity.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:110]: I just retained the functionality of exclusion, as we still have a chance to apply retention which is applied per entry.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala]: We can also have a streamlined version of this method to avoid materializing all entries on initializing FileStreamSource, though I think there's the another problem we should solve (file stream source log should not have bunch of entries - think about other data sources) and once we fixed that issue it won't matter at all.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:259]: This would help when we introduce a new condition on exclusion of entries.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:272]: This new approach is to avoid reading the next compact file log, which materializes all entries into the file.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala:101]: While I just keep this, I think we should just remove this. As I left TODO below, it hasn't been used, exists hypothetically.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala:108]: Note that this would also help to reduce the size of each entry. OK to leave only ADD_ACTION if JSON serializer/deserializer complains.

**Comment by HeartSaVioR** [comment]
cc. @tdas @zsxwing @jose-torres @gaborgsomogyi

**Comment by HeartSaVioR** [comment]
also cc. @xuanyuanking @uncleGen as they have been reviewed sibling PRs.

**Comment by HeartSaVioR** [comment]
Just to be sure, the test app with patch now writes the version 1589, which the log file size is 2.9G, with RES 1.025g.

**Comment by HeartSaVioR** [comment]
Combining efforts together:

* master branch + this patch (SPARK-30462) = batch 919 in 2 hours (120 mins)
* SPARK-30946 + this patch (SPARK-30462) = batch 919 in 11 mins

10x speedup, reduction of log file size (missed the batch 919 but 1369.compact only took 678M), consistent memory usage (under 1.2G)

**Comment by HeartSaVioR** [comment]
I've been running sustain tests (still running) and here's some observation:

- SPARK-30946 + SPARK-30462 just wrote compact batch 7309 (73,100,000 entries) which size is 3.5G, and RES has been less than 2G. It took 169517 ms (2.8 mins) to build a compact batch 7309, which is similar to build a compact batch 479 (4,800,000) without SPARK-30946. (15x processed)
- SPARK-30462 itself just wrote compact batch 1739 (17,390,000 entries) which size is 3.1G, and RES has been less than 2G.

**Comment by xuanyuanking** [comment]
Very impressive! I'll review this in 2 days.

**Comment by HeartSaVioR** [comment]
UPDATE: now SPARK-30946 + SPARK-30462 writes 11879 which RES is still less than 2G (around 1.7G). I'll stop the sustaining test for enough heap and run the another sustaining test for smaller heap (1.5G).

**Comment by HeartSaVioR** [comment]
UPDATE: SPARK-30946 + SPARK-30462 with lower down driver memory to 1.5G now writes batch 9039 which RES is around 1.3g. I guess the process uses up available memory if possible, but not leads to OOME.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala:177]: nit: duplicated code with `get`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala:203]: ditto, duplicate code with writeBatchToFile

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:110]: `compactLogs` requires to pass "all entries" which should be loaded into memory, right?

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:110]: The UT I had to remove leverages this behavior - it's only possible when we collect all entries, as it assumes DELETE_ACTION can come after ADD_ACTION "in any further batches" and compactLogs should be able to filter out. This also requires us to materialize all entries.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:184]: I had a caller leveraging this, but while working it seems to be removed. Let's make it simple.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala:101]: Just think out loud, DELETE_ACTION is not needed even we consider data compaction, if we do data compaction in compact batch. Nevertheless, let's focus on current state for now. DELETE_ACTION can be dropped.

**Comment by HeartSaVioR** [comment]
org.apache.spark.sql.hive.thriftserver.HiveSessionImplSuite.(It is not a test it is a sbt.testing.SuiteSelector)

This seems to be failing frequently. I'll see other build result being run via 124719.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala:101]: I'll just remove this condition - I believe we shouldn't concern about DELETE_ACTION, but even we concern about that, it would work better if we leave it to the reader, as FileStreamSinkLog won't be able to match pair of ADD_ACTION and DELETE_ACTION and remove the entries. 

Again, DELETE_ACTION should be just dropped, unless we have a plan to implement it properly.

File: sql/core/src/test/resources/structured-streaming/file-sink-log-version-2.1.0/8
Comment (HeartSaVioR) [inline sql/core/src/test/resources/structured-streaming/file-sink-log-version-2.1.0/8:3]: This should be removed - this gave false information as we support DELETE_ACTION previously, but in reality it was never implemented.

**Comment by HeartSaVioR** [comment]
```
Test build #124973 has finished for PR 28904 at commit 11cb693.

This patch fails Spark unit tests.
This patch merges cleanly.
This patch adds no public classes.
```

Known Kafka flaky tests.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by xuanyuanking** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala:100]: How about we still keep the `DELETE_ACTION` here?
```
if (log.action == FileStreamSinkLog.DELETE_ACTION) {
  false
} else {
  true
}
```
I also noticed that it's not implemented now, but it seems the only way to delete logs when compacting. I think it's also the original design goals. cc @zsxwing who is the author for SPARK-14678

**Review by xuanyuanking — COMMENTED** [review]
Only one comment for the DELETE_ACTION. Other changes LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala:100]: Please note that after the patch we are no longer able to match ADD_ACTION and DELETE_ACTION in compaction, so that's the responsibility of "reader" to exclude entities which have both ADD_ACTION and DELETE_ACTION, even we take DELETE_ACTION into account.

Returning `false` for DELETE_ACTION here simply drops the entities for DELETE_ACTION without dropping ADD_ACTION, hence some files would be resurrected on the reader side after compaction, making problems.

DELETE_ACTION is introduced years ago and we still haven't touched it. I think we should just defer and shouldn't deal with the thing we don't have and also don't have a clear plan to address. Once we design the new behavior, we would know what we should do. Dealing with something according to the imagination blocks us to move forward.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala:100]: Make sense. I also did some investigation and searched other thrid-party libraries, seems the DELETE_ACTION is not used. I think it's OK to design new behavior when we need it.

**Comment by xuanyuanking** [comment]
During searching the usage of `DELETE_ACTION`, I also found some usage for the thrid-party libraries of CompatibleFileStreamLog, e.g. elasticsearch-hadoop:
https://github.com/elastic/elasticsearch-hadoop/blob/9650205ee3732ee357eb04d1bb8a42a91a544a63/spark/sql-20/src/main/scala/org/elasticsearch/spark/sql/streaming/EsSinkMetadataLog.scala#L39

How about we still keep the `compactLogs` for not breaking the existing usage? I know it's not Spark API, but maybe it's better if we can implement this improvement with keeping this. WDYT?

**Comment by HeartSaVioR** [comment]
Well, I guess I already explained why `compactLogs` is the culprit of the memory issue, right? (https://github.com/apache/spark/pull/28904#discussion_r448049735)

All entries are materialized while compacting, and one of reasons to do so was providing these entries into `compactLogs`. In other words, if we don't provide all entries into compactLogs in any way to reduce memory usage, it's already breaking the existing behavior of `compactLogs`, and it's worse to change the semantic without any notice (if it's a public API... I'm not sure this is productive discussion as we all know it's private API).

**Comment by xuanyuanking** [comment]
`Well, I guess I already explained why compactLogs is the culprit of the memory issue, right? (#28904 (comment))`

Yep that's right. I'm also looking at the code in detail and try to find a way both keep this API and have the improvement. If it's hard to achieve, of course the improvement has a higher priority. I'll take a closer look today.

**Comment by HeartSaVioR** [comment]
Technically it's a private API, even not tagged as developer API - that said, it doesn't break anything in Spark's perspective. If we have confusion with availability of `org.apache.spark.sql.execution` package outside of Spark, then I'd rather say we may need to reconsider adding `private[execution]` on everywhere in the package.

**Comment by HeartSaVioR** [comment]
retest this, please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:110]: I take some time looking into the detail. Seems like we can use `Stream` here, which can delay the evaluation for each file status. Then we can both keep the API and have the improvement together. But for using `Stream[Int]` here, we'll have a restriction for the implementation of `compactLogs`, which also needs to return `Stream[Int]`, otherwise, the improvement of streamline will not take effect.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:110]: Is there a specific reason to avoid changing the method which is private API? I'd say it'd be less clear to provide the intention on downstream (technically we shouldn't concern with downstream though) - if you don't change the method signature but just change the semantic, no one would know about the change and it will be no-op for downstream.

I'll consider it if you have strong reason to do so, otherwise let's hear other's voice as well. cc. @zsxwing

Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:110]: Maybe not strong enough. IMO if we are able to avoid breaking the third-party libs, we'd better do so. Yep, let's hear more voice.
Great thanks for your patience and reply. It's the last concern left, other changes for this PR LGTM. I do think we need this improvement!

**Comment by HeartSaVioR** [comment]
cc. @tdas @zsxwing @jose-torres as it's technically not possible to merge only from @xuanyuanking 's review.

**Comment by itsvikramagr** [comment]
@HeartSaVioR - This is a much-needed fix. Thanks for it.

I have an orthogonal question. Why do we need to worry about compacting the file sink metadata? I can think of following reasons 
-  the downstream read operations can read the compacted metadata file to list all committed files. So they can avoid the listing cost and also improve performance
- Helps in exactly-once semantics. On task failures, we don't have to worry about deleting any files written. 

If the compacted metadata file size is running into GBs, the number of valid files would be in millions. In practice, the end-user will consider this sink path as a staging location and have another job to compact these small files into a final destination. 

for exactly-once semantics, we can add make changes in ManifestFileCommitter to delete files in the abort function. Or we can come up with some other alternatives.
 
In short, if we provide an option just to have last few commits in sink metadata to ensure SS is not impacted. And make changes in various readers not to read using metadata log files. Won't it help in ensuring the reliability of the streaming job?

**Comment by HeartSaVioR** [comment]
> for exactly-once semantics, we can add make changes in ManifestFileCommitter to delete files in the abort function. Or we can come up with some other alternatives.

I already provided the change (Spark 3.0.0 if I remember correctly), but this is only "best-effort". You cannot deal with crashing scenario.

Given we directly write the file to the final path, worth noting that reading output files in directory directly doesn't only mean you're reading duplicated outputs (at least once). This also means there's a chance you may be reading incomplete/corrupted files as well, in some sort of crashing during write.

Stepping back to explain the rationalization and the goal - providing holistic solution is not a goal. 

There're already lots of efforts being made, though these efforts are happening "outside" of Spark codebase, Delta Lake, Apache Iceberg, Apache Hudi, and probably more. I'd just reinvent the wheel if I try to address the entire problems, which I can't persuade anyone to provide me enough time to work.
(Please refer the comment which is the feedback I got - https://github.com/apache/spark/pull/27694#issuecomment-651454246)

My goal of the overall improvements in file stream souce/sink is, enabling end users run the query a bit longer without weird issues (like OOM). I just had to take easier (and limited) approach to solve the issue. e.g. regarding growing entries issue, while alternatives support data compaction to reduce the overall number of files, I proposed retention of the output files so that older files can be excluded in metadata log.

**Comment by HeartSaVioR** [comment]
cc. @tdas @zsxwing @jose-torres kindly reminder

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by dongjoon-hyun — COMMENTED** [review]
For the following notable change, I also agree with @HeartSaVioR 's assessment (https://github.com/apache/spark/pull/28904#discussion_r456729588) but didn't receive any strong feedback here.

```scala
- def compactLogs(logs: Seq[T]): Seq[T]
+ def shouldRetain(log: T): Boolean
```

Please let us know if you disagree, @tdas , @zsxwing , @jose-torres . If there is no written feedback, this is considered as OK in the community because two committers (@HeartSaVioR and me) are already agreeing in the written discussion.

For the remaining part, we will continue to discuss.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala:105]: Could you file a JIRA and make it IDed TODO, @HeartSaVioR ?
That will improve the visibility on `DELETE_ACTION` related issues.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:182]: Shall we mention `IllegalStateException` together explicitly?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:184]: Shall we mention `IllegalStateException` explicitly because the caller should know what kind of exception it is?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:190]: ditto. Shall we mention `IllegalStateException`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:110]: @HeartSaVioR . Can we have `def shouldRetain(log: T): Boolean = true` to simplify?
`true` is reasonable for the default value for `Log`, isn't it? And, in this PR, all derived class overrides with `true` only.

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for this improvement. I left a few minor comments, @HeartSaVioR .

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:185]: If the caller does not handle the exception properly, what is the consequence? Do we have a test case to cover it?

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:185]: How to ensure all the callers handle it properly? Do we need to introduce a test to cover it?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:133]: private?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:137]: private?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala
Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala:105]: Add a JIRA number in the comment?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala:161]: Nit: exception -> an exception

**Comment by HeartSaVioR** [comment]
Thanks for the review comments. Please allow some more days (probably next week) to address as I'm on vacation this week and have limited time on looking into laptop. :) Thanks again!

**Comment by dongjoon-hyun** [comment]
Sure. Have a nice vacation and take care, @HeartSaVioR .

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:110]: We'll enable the condition on file stream source and sink via #28363 and #28422 so eventually majority of derived classes will override the method. But I also agree `true` is the reasonable default value. I'll update this. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:184]: `IllegalStateException` is not the only exception this method can throw. It can also throw `IOException` as it deals with file, and also depends on `deserializeEntry`.

I think the type of exception won't affect what caller should do. Caller may want to deal with FNFE specifically, but still do the same for others, and even for FNFE it's safer to follow the guide on NOTE.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:185]: I'm not sure how this class ensures callers are following the guide. Did you mean we'd like to test this behavior with derived classes (file stream source/sink) log? Or we'd like to test this behavior with test-purpose implementation of CompactibleFileStreamLog?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala:190]: ditto. Would we want to mention all possible exceptions per method? `IllegalStateException` is the explicit one, but there're also many possible exceptions, starting from `IOException`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala:105]: Thanks for the suggestion. Will file an issue and add JIRA number into the code comment.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSinkLog.scala:106]: Let's consider only when we are open to introduce another version. Don't want to deal with compatibility even without bumping version.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @HeartSaVioR and all.
Merged to master for Apache Spark 3.1.0 on December 2020.

**Comment by HeartSaVioR** [comment]
Thanks all for reviewing and merging!

**Pull request closed without merging**
